streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -141,16 +144,17 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
 
         topology.addProcessor(name, new KStreamMap<>(mapper), this.name);
 
-        return new KStreamImpl<>(topology, name, null);
+        return new KStreamImpl<>(topology, name, sourceNodes, true);
```

Comment:
If we don't pass along the sourceNodes we can't ensure the topics are joinable and co-partition as we have no link to the original topics.
In the original code this would never have worked if you did:
``stream.map(..).join(..)`` as the mapped KStream would have a null sourceNode - it would have resulted in a ToplologyBuilderException